[nativert] add memory overlap debug assertion

dolpm opened the pull request:
Summary: better safe than sorry. will throw if memory overlap detected when using planned tensors and debug mode is enabled -- this will make our planning unit tests more robust.

Test Plan:
ci

Rollback Plan:

Differential Revision: D77327841

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -109,15 +126,71 @@ void AliasAnalyzer::create_or_update_lifetime(const Value* value, size_t i) {
+
+  // we want aliases_[x] to return the value from which x
+  // inherits its dataptr (there can only be 1).
+  // as such, we want to all values that do not meet this
```
want to add?

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -96,7 +113,7 @@ void AliasAnalyzer::maybe_update_aliases_from_schema(
                 << " may contain input/output alias: " << input->id() << " -> "
                 << output->id();
-        aliases_[output].insert(input);
+        aliases[output].emplace(input);
```
I have asked you before, but I don't remember now, and I didn't find definition of  
c10::FastMap<const Value*, c10::FastSet<const Value*>> anywhere in this file. 

say, you have c = op(a, b), when are you going to have both a and b an alias of c?

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.h:
```diff
@@ -70,16 +83,41 @@ class AliasAnalyzer {
+  // source (i.e., the value that owns its
+  // associated dataptr)
+  void squash_deep_aliases(
+      c10::FastMap<const Value*, c10::FastSet<const Value*>> aliases,
```
I am confused... 

I can understand alias is pair-wise relationship, that's why you have private member
c10::FastMap<const Value*, const Value*> aliases_;

but why the alias here have a different type! what's the inner FastSet for? 
c10::FastMap<const Value*, c10::FastSet<const Value*>> aliases,

dolpm replied:
i responded to why it is a set in the comment above, but this is indeed an issue.

when i implemented the squashing, i forgot about the ambiguous cases (e.g., no schema). so even though in reality a tensor can only alias one source at a time, it may _be able_ to alias multiple. thus, the squashed datastructure needs to be a map from the alias to _the set of all possible sources_. in most cases, this will be a singleton set, but there are a few cases where it might not be, so we gotta support it.

dolpm commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +160,123 @@ void LayoutManager::populate_tensor_values() {
+  const auto& alive_values = alias_analyzer.alive_values_at_time(t);
+
+  // make sure active memory intervals are non-overlapping
+  // by sorting them by start, and ensuring
```
on comparison of pairs: https://cplusplus.com/reference/utility/pair/operators/

i will add this to the comment.

dolpm commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +160,123 @@ void LayoutManager::populate_tensor_values() {
+      TORCH_CHECK_LE(t, lt.end);
+
+      // prevent recomputation from occuring
+      c10::FastSet<ValueId> checked_values;
```
right

SherlockNoMad commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +160,123 @@ void LayoutManager::populate_tensor_values() {
+        continue;
+      }
+
+      const auto& src_iv = parent_frame_.getIValue(src_ptr->id());
```
checking logic looks duplicated for iv and src_iv? 

refactor them into a function?

dolpm commented on torch/nativert/executor/memory/LayoutManager.h:
```diff
@@ -178,6 +211,9 @@ class LayoutManager {
 
   std::vector<const at::Tensor*> planned_tensors_;
   std::vector<size_t> planned_tensors_max_nbytes_local_;
+#ifndef NDEBUG
```
iiuc this class member won't be optimized out of the structure, even if it isn't used...

dolpm commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -109,15 +122,52 @@ void AliasAnalyzer::create_or_update_lifetime(const Value* value, size_t i) {
 }
 
-void AliasAnalyzer::maybe_extend_lifetimes(const Graph& graph) {
+void AliasAnalyzer::squash_deep_aliases(
```
the existing alias analyzer unit tests will cover this since it influences an aliases' _source_ value

SherlockNoMad commented on torch/nativert/executor/memory/LayoutPlanner.h:
```diff
@@ -111,6 +120,9 @@ class LayoutPlanner {
   std::vector<ValueId> planned_values_;
   std::vector<AllocationSpec> planned_allocation_specs_;
   std::vector<std::atomic_size_t> planned_values_historical_max_nbytes_;
+  // alive_values_at_time_[x] will return the set of values that are
+  // alive/allocated during the execution of graph.nodes[x]
```
nice and clear!

SherlockNoMad commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+      // NOTE: it's possible the source is ambiguous, hence
+      // why get_sources_of_alias returns a set (although it's usually a
+      // singleton set)
+      if (const auto* srcs_of_v = alias_analyzer.get_sources_of_alias(v);
```
I like srcs_of_v, far better than src_prt

SherlockNoMad commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+      // we want to try to get the source (i.e., the value)
+      // that actually owns the storage
+      //
+      // NOTE: it's possible the source is ambiguous, hence
```
let's chat offline about the source is ambiguous situation. 
non blocking for this PR tho

SherlockNoMad replied:
just an idea. 
aliasing relationship should be know after the first forward run.

dolpm commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+      if (const auto start = layout_buffer_.get_offset_from_ptr(
+              storage_impl->data_ptr().get());
+          start.has_value()) {
+        return std::make_pair(*start, *start + storage_nbytes - 1);
```
yes, we need.

start + storage_nbytes would be exclusive

e.g., 

32 + 64 --> 96
                      ^
                      byte 96 in the buffer is not actually taken since the start is inclusive and thus is included in the size.
                      
                      the interval we want is [32, 95] which is 64 bytes inclusive
                      
the reason we compare  with <= on L290 is because we need to ensure the alias is a sub-slice of the source

e.g.,

|######1######|
    |#####2#####|
                             ^
                             s2 would be a valid alias of s1 here.


|######1######|
|######2######|
                             ^
                             same here.

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.h:
```diff
@@ -70,16 +84,41 @@ class AliasAnalyzer {
   c10::FastSet<const Value*> values_associated_with_outputs_;
+  // alive_values_at_time_[i] = values that are "alive" during the
+  // computation of node i
+  std::vector<std::vector<const Value*>> alive_values_at_time_;
```
Just a reminder

It's accidental that in current sequential executor, node_id match the execution "time". 

Execution order might not follow graph node order, if we do some execution order optimization.

dolpm replied:
yes. unfortunately a ton of code depends on this invariant -- if execution ordering is not well defined (and/or nondeterministic) we will have major issues across the board. this is the main reason why inter-op won't work with memory planning.

dolpm replied:
the edge case here is there is some optimization that can only be reasoned at execution time (e.g., based on batch dim). can't think of one off the top of my head rn, though.

SherlockNoMad commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+  auto it = intervals.begin();
+  size_t prev_end = it->second;
+  while (++it != intervals.end()) {
+    TORCH_CHECK_LT(prev_end, it->first /* cur_start */);
```
also, LT is correct here, since prev_end has been offset by 1.

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -16,27 +19,40 @@ AliasAnalyzer::AliasAnalyzer(
+
+  for (const auto& [v, lifetime] : lifetimes_) {
+    for (const auto t : c10::irange(lifetime.start, lifetime.end + 1)) {
+      alive_values_at_time_[t].emplace_back(v);
```
this object would be pretty big. 
should I be worried about it's memory usage? 
also, is this for debug only?

dolpm replied:
let me make this debug only - i agree this will be quite large. good catch.

dolpm commented on torch/nativert/executor/memory/LayoutManager.cpp:
```diff
@@ -157,6 +161,165 @@ void LayoutManager::populate_tensor_values() {
+    //
+    // if the checks pass, add the interval [alloc_start, alloc_start +
+    // alloc_nbytes] to the set of intervals
+    auto check_allocation_bounds =
```
yeah this comment should be [alloc_start, alloc_start + alloc_nbytes) now though

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -109,15 +125,74 @@ void AliasAnalyzer::create_or_update_lifetime(const Value* value, size_t i) {
+    c10::FastMap<const Value*, c10::FastSet<const Value*>> aliases,
+    const Graph& graph) {
+  for (auto& node : graph.nodes()) {
+    for (const auto& input : node.inputs()) {
```
Is it better to iterate over node's output value?
one value could be consume by multiple node's inputs, but one value can only by produced by one node.

SherlockNoMad replied:
Does the visiting order of values matters? 
If not, I think Graph class is missing a util class to iterate over graph's values.

dolpm replied:
> Does the visiting order of values matters?

yeah. when we visit a node we need to ensure that all possible srcs of the nodes' associated aliases (if they exist) have already been squashed

dolpm replied:
i think iterating over outputs should be OK tho.

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -109,15 +125,74 @@ void AliasAnalyzer::create_or_update_lifetime(const Value* value, size_t i) {
+        aliases_[alias].emplace(src);
+      }
+    }
+  }
```
not a big fan of hiding aliases_ initialization inside a unrelated function... 

this should be a standalone function

dolpm replied:
agree

SherlockNoMad commented on torch/nativert/executor/memory/AliasAnalyzer.cpp:
```diff
@@ -129,10 +204,11 @@ void AliasAnalyzer::maybe_extend_lifetimes(const Graph& graph) {
 
-          if (eol == graph.nodes().size() - 1 /* aliases output */) {
-            values_associated_with_outputs_.insert(src);
+          if (aliases_.find(src) == aliases_.end() &&
```
for sanity, please rename aliases_ to finalized_aliases_ or something to avoid miss use. 

TBH, it's an anti-pattern to have both `aliases_` and local variable `aliases`.

dolpm replied:
yeah, i updated to mutate aliases_ directly everywhere. i had the two copies back when they had two different structures (aliases_ didn't have the set)....

SherlockNoMad reviewed (approved):
Thanks for addressing all the issues, this looks great now. 

I was not supposed to scrutinize a diff about debug assertion, but I wanna use this diff to establish some kinda of protocols between us. This will help our productivity in the future. 

- For large changes, having a detailed summary is really helpful  for review to grasp what you are trying to do. 
- For complicated algorithms, try to break it into smaller utils functions for better readability
- If you show good test coverage, reviewer usually have higher confidence for fast approving. 
- Code comments for complicated logics plz. Your future self will also thank you for this. 

Final tip, review your own code. I often catch trivial bugs when I just re-read my own code with a reviewer hat.

dolpm committed:
[nativert] add memory overlap debug assertion (#157290)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/157290

better safe than sorry. will throw if memory overlap detected when using planned tensors and debug mode is enabled.

Test Plan:
ci

Rollback Plan:

Reviewed By: SherlockNoMad, zhxchen17

Differential Revision: D77327841

huydhn commented:
@pytorchbot merge

huydhn commented:
@pytorchbot merge -i
